### DISCUSSION TITLE
How can I retrieve the peer certificate?

### DISCUSSION THREAD

**Discussion by Xromern** [actix/actix-web#3458] [Q&A] [answered]
The data from the peer certificate is used for authentication.

For example, in Node.js, I can retrieve it like this:

```
this.router.get('/data', async (req, res) => {
    const cert = req.socket.getPeerCertificate();
});
```
I tried accessing the TlsStream in middleware using bind_rustls_0_23 , but I couldn't get it to work.

I would appreciate any help.

**Comment by robjtede** [comment, marked as answer]
An example of how to read client certificates is shown here:

https://github.com/actix/examples/blob/0523eea2f6a8a0fad66d0fbac2e067f7a0a137c6/https-tls/rustls-client-cert/src/main.rs#L30-L70

...using an `on_connect` function which adds to the connection-level data.

**Reply by Xromern** [reply]
Thank you so much, I searched everywhere but missed this example!
